### PR TITLE
parser: fix lexer that treat 9eTSs as a float

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Scan 9eTSs should get an identifier token

### What is changed and how it works?

Fix https://github.com/pingcap/tidb/issues/8983

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @winoros @lysu

File: lexer_test.go [hunk 1]
```diff
@@ -232,6 +233,8 @@ func (s *testLexerSuite) TestIdentifier(c *C) {
 		{"1_x", "1_x"},
 		{"0_x", "0_x"},
 		{replacementString, replacementString},
+		{"9e", "9e"},
+		{"9eTSs", "9eTSs"},
```
Comment (lysu) [inline lexer_test.go]: maybe need a test-case for `9e9e` :shipit:

Comment (kennytm) [reply lexer_test.go]: `9e9e` is not recognized as an single identifier though.

```sql
mysql> select 1 9e9e;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '9e9e' at line 1

mysql> select 9e9e;
+------------+
| e          |
+------------+
| 9000000000 |
+------------+
1 row in set (0.00 sec)
```

By extension, could we also check the following:

```sql
mysql> select 1 0xaaahhh;
+----------+
| 0xaaahhh |
+----------+
|        1 |
+----------+
1 row in set (0.00 sec)

mysql> select 1 0b123;
+-------+
| 0b123 |
+-------+
|     1 |
+-------+
1 row in set (0.00 sec)

mysql> select 1 0x3p4;
+-------+
| 0x3p4 |
+-------+
|     1 |
+-------+
1 row in set (0.00 sec)
```

Comment (tiancaiamao) [reply lexer_test.go]: 9e9e = 9e9 + e @lysu

Comment (tiancaiamao) [reply lexer_test.go]:
```
mysql> select 9e9e;
+------------+
| e          |
+------------+
| 9000000000 |
+------------+
1 row in set (0.00 sec)

mysql> select 9e9 e;
+------------+
| e          |
+------------+
| 9000000000 |
+------------+
1 row in set (0.00 sec)
```

@kennytm

Comment (tiancaiamao) [reply lexer_test.go]: Test added @lysu

Comment (tiancaiamao) [reply lexer_test.go]:
```
mysql> create table t (0b123 int);
Query OK, 0 rows affected (0.19 sec)
```

You're right ... @kennytm 
0b123 is also an identifier!

**Comment by tiancaiamao** [comment]
The rule is weird...

```
0x7fz != 0x7f + z
0b12 != 0b1 + 2
9e9a = 9e9 + a
```

I have some updates, PTAL @kennytm @lysu

File: parser_test.go [hunk 2]
```diff
@@ -2040,7 +2040,7 @@ func (s *testParserSuite) TestType(c *C) {
 		// for bit
 		{"select 0b01, 0b0, b'11', B'11'", true, "SELECT b'1',b'0',b'11',b'11'"},
 		{"select 0B01", false, ""},
-		{"select 0b21", false, ""},
+		// {"select 0b21", false, ""}, 0b21 is an identifier, the statement could parse.
```
Comment (kennytm) [inline parser_test.go]: 🤔 `0B01` is also an identifier. I think both lines 2042 and 2043 should have the same treatment (either both can be parsed or both need to be commented out).

Comment (tiancaiamao) [reply parser_test.go]: Fixed

**Pull request merged**